[IV-23-24] Objetivo 4

shvtwp opened the pull request:
* [x]  ¿Hay un camino claro que permita ir desde el código hasta la historia de
   usuario que desarrolla, pasando por el mensaje de commit y el issue
   correspondiente?
* [x]  ¿Se ha intentado cubrir con tests una parte de la lógica de negocio que se
   esté desarrollando?
* [x]  ¿Hay un producto mínimamente viable que describa lo que se va a entregar en
   este objetivo?
* [x]  ¿En este producto mínimamente viable se han priorizado unas clases,
   módulos o paquetes más fundamentales frente a otros, movidos a otro
   PMV/Hito?
* [ ] ¿Estoy leyendo todo lo que marco o simplemente marco lo que se me pone por
delante?
* [x] ¿Se han documentado las elecciones de biblioteca de aserciones y del
   de *test runner*?
* [x] ¿Te has asegurado que lo que mencionas como biblioteca de aserciones y
test runners lo son realmente y tienen la misma funcionalidad? ¿O estás poniendo
`unittest` como *test runner* y `pytest` como biblioteca de aserciones?
* [ ] ¿Esa documentación incluye criterios de aceptación y también criterios de
búsqueda para las diferentes opciones?
* [x]  ¿Se han seguido los principios F.I.R.S.T. en la creación del test y
   se ha documentado cómo se ha hecho?
* [ ]  ¿Se describe claramente en el PMV los tests que hay que pasar para que se
   considere viable?
* [ ] ¿Se han testeado también las excepciones?

Ahora mismo solo está implementada la lógica de negocio del primer milestone

shvtwp committed:
Establece las herramientas test #21

closes #21

manuelvico0102 commented on docs/herramientas_tests.md:
```diff
@@ -0,0 +1,40 @@
+externos.
+También son cualidades a tener en cuenta la eficiencia y la asiduidad de
+mantenimiento, entre otras.
+
```
Hay algunos criterios que mencionas, pero luego en las diferentes opciones a elegir no las mencionas.
Además, es recomendable usar [Snyk Advisor](https://snyk.io/advisor/) para comparar.

shvtwp replied:
Solucionado en 27f436a

manuelvico0102 commented on docs/herramientas_tests.md:
```diff
@@ -0,0 +1,40 @@
+Deno ofrece su propio módulo de aserciones en la librería estándar [Deno_std](https://docs.deno.com/runtime/manual/basics/testing/assertions),
+pero también puede usarse la librería [Chai](https://deno.land/x/chai@v5.0.0-rc.0), pensada
+exclusivamente para Deno.
+
```
Creo que sería conveniente que comentases más opciones de librerías de aserciones, como algunas relacionadas con tu lenguaje seleccionado.

shvtwp replied:
Solucionado en 27f436a

manuelvico0102 commented on test/horario_test.ts:
```diff
@@ -0,0 +1,156 @@
+
+mod.describe ("Horario - Asignación de turnos", () => {
+
+    function todasLasHorasAsignadas(horario: Horario): boolean {
```
Las funciones que has declarado en el test pienso que es parte de tu lógica de negocio, entonces no creo que sea muy buena práctica declararlas en el archivo para el test

shvtwp replied:
Hola! Es que son funciones que no necesito para la lógica de negocio como tal, pero sí para testearla, entonces no me queda muy claro por qué colocarlas en, por ejemplo, horario.ts

PabloBarTo replied:
Buenas! Sinceramente yo creo que están bien esas funciones ahí, ya que si solo se usan para probar que funcionan los test se podría implementar donde están implementados. Sin embargo dejaría que el profesor le echara un vistazo por si el considera que no deberían de estar ahí y tienen que estar en otro lado.
No resuelvas esta conversación para que luego el profesor pueda revisarlo y cuando soluciones todo lo que te ha comentado @manuelvico0102, que creo que solo queda arreglar lo de los issues, mencionamos al profesor.

JJ replied:
Yo creo que la cuestión no es si deben estar en la lógica de negocio o en el test, sino si deben estar en este fichero o en otro fichero; más adelante decidir dónde exactamente debe estar ese fichero. En todo caso si se ha planteado la necesidad de escribir esas funciones, habrá tenido que ser como solución a un problema. Si el problema está relacionado directamente con la HU, es lógicamente parte de la lógica de negocio; si sólo son funciones auxiliares para los tests, y por tanto sólo indirectamente relacionados con la lógica de negocio , lo lógico es que estén en el directorio de test; pero en cualquiera de los casos parece más conveniente, por modularización, que estén aparte.

shvtwp replied:
Solucionado en a0cecb3

manuelvico0102 reviewed (changes requested):
Buenas, voy a ser tu revisor para este objetivo. También te recomiendo que deberías crear más issues para ir indicando lo que vas haciendo y cuando hagas un commit hacer referencia a ese issue en vez de a la historia de usuario directamente.

shvtwp commented:
@JJ listo para revisión.

JJ commented on docs/herramientas_tests.md:
```diff
@@ -0,0 +1,51 @@
+Su versión más reciente se lanzó el 12 de diciembre de 2023, y se
+actualiza aproximadamente cada tres meses.
+
+Usaré la opción ofrecida por la librería estándar de Deno para no tener
```
REalmente no hay tanta diferencia en Deno entre unas y otras; deno no lo incluye en su distribución por defecto y tienen que bajarse igualmente y cachearse. Puede ofrecerte una pequeña ventaja de mantenimiento, pero otras opciones si están bien actualizadas pueden ser también muy interesantes.

shvtwp replied:
Solucionado en 1d98d5d

JJ commented on drakefile.ts:
```diff
@@ -12,6 +12,12 @@ task("ejecutarAplicacion", [], function ejecutarAplicacion() {
+//Tarea para ejecutar los tests
+desc("Ejecutar los tests");
+task("test", [], function ejecutarTests() {
+    sh("deno test test/horario_test.ts");
```
¿Has probado a no poner explícitamente el nombre del test? Más adelante tendrás que hacer más, posiblemente.

JJ replied:
Por favor, resuelve todos los comentarios con cambios en la línea o en otras líneas indicando el commit en el que se ha resuelto.

shvtwp replied:
Solucionado en 60813f1

JJ commented on lib/cuidador.ts:
```diff
@@ -0,0 +1,33 @@
+      public horasCuidado : number
+   ){}
+
+   public getNombre(): string {
```
Ya en el [objetivo 2 se señaló que se usaran sólo los getters y setters que fueran necesarios](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-2.md#sobre-el-c%C3%B3digo-definido). Además, TS/JS tienen una forma específica para definir getters, si fuera necesario hacerlo.

shvtwp replied:
Solucionado en fd73353

JJ commented on lib/cuidador.ts:
```diff
@@ -0,0 +1,33 @@
+      return this.tipo;
+   }
+
+   public setHorasOcupadas(hora: number, duracion: number){
```
Los setters son aún más discutibles que los getters. Por varias razones
1. La clase se llama "cuidador" y le estás encargando que incluya en su estado las horas en las que está ocupado/a. Las clases deben tener un solo cometido, y pro el nombre parece que se trata de los datos generales del cuidador.
2. Por otro lado, también estás haciendo un diseño en el que conviertes cuidador en entidad, al ser mutable; en vez de gestionar la entidad las horas ocupadas, es la entidad la que tiene que llamar al cuidador para establecer las horas ocupadas. Esto puede causar problemas por la indirección, pero también por el control del estado.

shvtwp replied:
No estoy segura de haberlo entendido bien, lo he modificado pero me gustaría aclararlo en clase

shvtwp replied:
Solucionado en c269cbd

JJ commented on lib/cuidador.ts:
```diff
@@ -0,0 +1,33 @@
+   
+   constructor(
+      private nombre: string,
+      public horasOcupadas: Array<boolean> = new Array(168).fill(false),
```
El problema con esta línea modificada en fe94d2c es que estás modificando directamente el código apuntando a la historia de usuario #3. En general, hay que seguir el mismo procedimiento que en el objetivo 2: crear issues que representen una serie de problemas, y a partir de ahí desarrollar el código. La HU normalmente expresa el problema desde el punto de vista del cliente, y es imposible deducir directamente del mismo la lógica de negocio: hay que seguir un procedimiento, una vez más *domain driven design*, describiendo los diferentes problemas y creando las estructuras de datos y funcionalidad necesaria.

shvtwp replied:
Solucionado en fd73353

JJ commented on lib/cuidador.ts:
```diff
@@ -0,0 +1,33 @@
+   }
+
+   public setHorasOcupadas(hora: number, duracion: number){
+      if (hora + duracion > 168){
```
Es una buena práctica usar siempre constantes con nombre, en vez de literales. Adicionalmente, esta línea la has modificado en el mismo ticket indicado arriba, y sólo apunta a la historia de usuario, con lo que es complicado o imposible, para quien vea el código, entender si es correcto o no.

shvtwp replied:
Solucionado en 87a69c3

JJ reviewed (changes requested):
Gracias a @manuelvico0102 y @PabloBarTo por la retroalimentación y gracias a @shvtwp por actuar tan rápidamente; pero me temo que en este caso es necesario dar un paso atrás y seguir las buenas prácticas: sólo se desarrolla código para resolver problemas que estén bien expresados en issues, porque es la única forma de garantizar que se cubren las necesidades del usuario. Por favor, hazlo así.

shvtwp committed:
Elige las herramientas para test closes #21

shvtwp committed:
Renombra familiar_cuidador.ts a cuidador.ts

shvtwp committed:
Modifica representación de cuidador

Es necesario indicar el tipo del cuidador, no solo si es adulto o es joven. Además, se ha cambiado la representación de las horas ocupadas de un cuidador; pasan de ser turnos a horas ocupadas o libres #28

shvtwp committed:
Elimina números mágicos en cuidador.ts #28

shvtwp committed:
Modifica la representación de horario.ts #23

Además, establece el límite de horas semanales de cada tipo de cuidador. Las horas que queden sin asignar serán las horas que deberán ser cubiertas por un profesional

shvtwp committed:
Añade asignación de horas ocupadas #28

Se añade en horario.ts una función que permite asignar a cada cuidador sus horas ocupadas a modo de turno, indicando la hora en la que empieza a estar ocupado y cuántas horas más seguirá estándolo

shvtwp committed:
Implementa test todasLasHorasAsignadas #26

Por una parte, se implementa el test que se va a ejecutar en el archivo horario_test.ts; por otro lado, en el archivo test_aux.ts se implementan las funciones auxiliares necesarias para la ejecución del test closes #26

shvtwp committed:
Implementa test turno sobrepasado #25

Este test se asegura de que ningún cuidador sobrepase el cupo de horas establecido para él closes #25

shvtwp committed:
Implementa test incompatibilidad horaria #27

Este test se asegura de que no se de el caso de que un cuidador ocupado a cierta hora sea asignado para cuidar de la persona en situación de dependencia en ese mismo momento closes #27 . Con esto se completan los test previstos closes #24

shvtwp committed:
Implementa asignación horaria a cuidador #23

Este commit implementa la lógica de negocio necesaria para abordar la HU1 #3. Se añade un algoritmo que asigna las horas de cuidado al cuidador que esté disponible en ese momento, priorizando más o menos según el tipo de persona que sea. En la función compararPorPrioridad se refleja el orden de la misma closes #23

shvtwp committed:
Añade test a Drakefile #21

shvtwp committed:
Actualiza iv.yaml #21

shvtwp committed:
Actualiza README.md #21

Incluye el comando para ejecutar los test

shvtwp commented:
@JJ Listo para revisión

JJ reviewed (commented):
Queda un comentario por resolver; le doy un vistazo sobre la marcha cuando lo hagas.

JJ reviewed (approved):
👍
